### PR TITLE
User can sign up

### PR DESCRIPTION
**Pull request by koggs**
#### What I did:
- Add gems Sorcery and validates email format of
- Create form for user authentication
- Create user controller and model

### CODE REVIEW

File: app/controllers/users_controller.rb [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+    @user = User.new(user_params)
+    if @user.save
+      login(params[:email], params[:password])
+      flash[:success] = "Welcome! You have signed up successfully"
+      redirect_to root_path
+    else
+      if @user.errors.messages[:password] != []
+        flash[:error] = "Password has to be minimum 6 caracters"
+      elsif @user.errors.messages[:password_confirmation]
+        flash[:error] = "Account not created, passwords don't match"
```
Comment (juliaramstedt) [inline app/controllers/users_controller.rb]: Why a flash for these password errors- isnt' it enough to display them along with other model errors on the edit-view?

File: app/controllers/users_controller.rb [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+  end
+
+  private
+
+  def user_params
+    params.permit(:email, :password, :password_confirmation)
+  end
+end
+
+#  case
```
Comment (juliaramstedt) [inline app/controllers/users_controller.rb]: Is this comment supposed to be here?

File: Gemfile [hunk 3]
```diff
@@ -16,6 +16,8 @@ gem "jbuilder", "~> 2.5"
 gem "bootsnap", ">= 1.1.0", require: false
 gem "bootstrap", "~> 4.0.0"
 gem "jquery-rails"
+gem "sorcery"
```
Comment (juliaramstedt) [inline Gemfile]: it's a good idea to specify the version of your gems, e.g. `gem "sorcery", "~> 0.13.0" `

File: Gemfile [hunk 4]
```diff
@@ -16,6 +16,8 @@ gem "jbuilder", "~> 2.5"
...
+gem "validates_email_format_of"
```
Comment (juliaramstedt) [inline Gemfile]: I wouldn't use this gem, it's not very maintained. Consider using a regex instead:

https://guides.rubyonrails.org/active_record_validations.html#format

**Review by juliaramstedt — CHANGES_REQUESTED** [review]

File: app/controllers/users_controller.rb (hunk 2 above)
Comment (koggs) [reply app/controllers/users_controller.rb]: No I forgot to delete it :D

File: Gemfile (hunk 3 above)
Comment (koggs) [reply Gemfile]: Done :)

File: app/controllers/users_controller.rb (hunk 1 above)
Comment (koggs) [reply app/controllers/users_controller.rb]: done :)

File: Gemfile (hunk 4 above)
Comment (koggs) [reply Gemfile]: I removed te gem and used regex. I used the gem because people were saying it was much simpler :)

File: app/models/user.rb [hunk 5]
```diff
@@ -0,0 +1,11 @@
+# frozen_string_literal: true
+
+class User < ApplicationRecord
+  authenticates_with_sorcery!
+
+  VALID_EMAIL_REGEX = /\A[\w+\-.]+@[a-z\d\-.]+\.[a-z]+\z/i.freeze
```
Comment (juliaramstedt) [inline app/models/user.rb:6]: you shouldn't have to run `.freeze` on it, but it's ok!

File: app/controllers/users_controller.rb [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+  end
+
+  def create
+    @user = User.new(user_params)
+    if @user.save
+      login(params[:email], params[:password])
+      flash[:success] = "Welcome! You have signed up successfully"
+      redirect_to root_path
+    else
+      flash[:error] = @user.errors.full_messages
```
Comment (juliaramstedt) [inline app/controllers/users_controller.rb]: Usually for validation errors, you don't display them in a flash message but instead render them on the page using something like: 

```
   <% if @user.errors.full_messages.any? %>
      <%= @user.errors.full_messages.join(", ") %>
   <% end %>
```

or you can even print the individual messages out next to their fields,

something like...
```
   <%= @user.errors["email"] %>
```

which is usually used to add an `.has-error` css class or similar.

But this is ok! Just thought you might want to be aware of it.

**Review by juliaramstedt — APPROVED** [review]

File: app/controllers/users_controller.rb (hunk 6 above)
Comment (koggs) [reply app/controllers/users_controller.rb]: I was aware but it wasn't working

**Pull request merged**